Gems/Archive/Code/Include/Archive/Clients/ArchiveInterfaceStructs.inl

```diff
@@ -283,5 +321,197 @@ namespace Archive
+    //! AlignedDownTo2MiB(4 MiB - 1) = 2 MiB
+    //! Next divide the value that was aligned down by 2 MiB = 1
+    //! This makes sure the block containing the start offset is read
+    //!
+    //! The last block to read would would be at index 4 using the following calculations
+    //! AlignedUpTo2MiB(4 MiB - 1 + 4 MiB + 2) == AlignedUpTo2MiB(8 MiB + 1) = 10 MiB
+    //! Next that aligned up value is divided by 2 MiB = 5
+    //! This makes sure the last block containing the the bytes to read is included before
+    //! the sentinel value
+    //!
+    //! This forms a range of [1, 5), therefore the blocks of 1, 2, 3, 4 are read
+    //! The final byte from block index 1 is read (running count = 1)
+    //! All bytes from block index 2 and 3 are read (running count = 1 + 2 MiB * 2 = 4 MiB + 1)
+    //! The first byte from block index 4 is read (final count = 4 MiB + 2)
+    //! As 4 MiB + 2 is the amount of bytes requested to be read from the user it is returned.
+    constexpr AZStd::pair<AZ::u64, AZ::u64> GetBlockRangeToRead(AZ::u64 uncompressedOffset, AZ::u64 bytesToRead)
+    {
+        // Index value that contains the first block to read for the file
+        AZ::u64 blockLineIndexForOffset = AZ_SIZE_ALIGN_DOWN(uncompressedOffset, ArchiveBlockSizeForCompression)
+            / ArchiveBlockSizeForCompression;
+
+        // Sentinel index value that is one above the last compressed block to read
+        AZ::u64 blockLineSentinelForBytesToRead = AZ_SIZE_ALIGN_UP(uncompressedOffset + bytesToRead, ArchiveBlockSizeForCompression)
+            / ArchiveBlockSizeForCompression;
+
+        return { blockLineIndexForOffset, blockLineSentinelForBytesToRead };
+    }
+
+    // Use an immediate invoked lambda expression to test the GetBlockRangeToRead
+    // condition at compile time
+    static_assert([]() {
+        auto blockRange = GetBlockRangeToRead(4_mib - 1, 4_mib + 2);
+        return blockRange.first == 1 && blockRange.second == 5;
+    }(), "A start offset of 4 MiB - 1 and a read size of 4 MiB + 2, should result in a block range of [1, 5)");
+
+
+    // Implementation of the GetBlockLineFromBlockIndex function
+    constexpr GetBlockLineIndexResult::operator bool() const
+    {
+        return m_errorMessage.empty();
+    }
+
+    constexpr GetBlockLineIndexResult GetBlockLineIndexFromBlockIndex(AZ::u64 blockCount, AZ::u64 blockIndex)
+    {
+        GetBlockLineIndexResult result;
+        if (blockIndex >= blockCount)
+        {
+            result.m_errorMessage = "Block index is out of range of the number of block count."
+                " A block line index cannot be returned";
+            return result;
+        }
+
+        auto GetFirstIndexForFinalBlockLineSetOf3 = [](AZ::u64 blockCount) constexpr
+        {
+            auto GetFinalIndexForFinalBlockLineSetOf3 = [](AZ::u64 blockCount) constexpr
+            {
+                return blockCount <= MaxBlocksNoJumpEntry
+                    ? MaxBlocksNoJumpEntry
```

Comment:
The `blockCount <= MaxBlocksNoJumpEntry` is checked before using this function in the lines below, so is this check necessary here?

Reply:
No it is not.
I'll remove in this PR if there is a need for a non-trivial update.

Reply:
Also make sure which logic is right. The one below returns zero, this one returns `MaxBlocksNoJumpEntry` if `blockCount <= MaxBlocksNoJumpEntry`

Reply:
Well the inner lambda condition is always returning the "else" part of the Ternary since the condition is checked before calling it.

The entire purpose of the two function is to return the minimum block index for the final set of block lines that don't have a jump entry (confusing and hard to explain, I know).

But pretty much every compressed file in the Archive that is >16 MiB starts with its first block line (index 0) with a jump entry.
And then every three block lines afterwards(index 3, 6, 9, ...) would have a Jump entry if there more than 18 MiB (9 blocks) remaining in the file from the start of that block line.

It is explained better in the Streaming Archive document at [here](https://github.com/o3de/sig-core/blob/decc359e5fbc3a718e193bb92e6d80e55e19fb4f/rfcs/rfc-core-2023-04-20-streaming-archive.md#block-offset-table-section), but the implementation is complicated.